### PR TITLE
build: update MegaLinter configs for PHP

### PR DESCRIPTION
**Pull request by eitsupi**
Part of #1994

Reflects the Lint setting from #2039

### CODE REVIEW

**Comment by eitsupi** [comment]
psalm (https://megalinter.io/latest/descriptors/php_psalm/) seems complaining about FFI.

```log
  ERROR: UndefinedClass - ../../../github/workspace/prql-php/src/Compiler.php:32:13 - Class, interface or enum named FFI does not exist (see https://psalm.dev/019)
      private \FFI $ffi;
```

I don't understand PHP at all and didn't think I could write a psalm config file, so can I disable it for now?
If anyone finds psalm useful, by all means add a config file and re-enable this.

**Comment by eitsupi** [comment]
@vanillajonathan Could you take a look at this?
I know nothing about PHP, so I would appreciate your opinion on whether this configuration is meaningful.

File: .mega-linter.yaml [hunk 1]
```diff
@@ -26,9 +27,8 @@ DISABLE_ERRORS_LINTERS:
   - JSON_JSONLINT
   - MAKEFILE_CHECKMAKE
   - MARKDOWN_MARKDOWN_LINK_CHECK
-  - PHP_PHPCS
-  - PHP_PHPSTAN
-  - PHP_PSALM
   - SPELL_MISSPELL
   - SQL_TSQLLINT
   - YAML_V8R
+PHP_PHPCS_ARGUMENTS: --standard=PSR12
```
Comment (eitsupi) [inline .mega-linter.yaml]: @nvuillam Sorry for tagging you. Is this the right way to set this?

I want to set like this

```sh
phpcs --standard=PSR12 src tests
```

I searched the documentation and on GitHub, but I'm not sure......

**Comment by eitsupi** [comment]
Surprisingly, phpstan generates the following error when inspecting only php files (d68f4c6abf5e30ef0ec5c94aaf431f94c4a90749) while not generating an error when the entire repository is subject to lint (3727c472444576858146c985edbafefe221fea10) !

```log
  - Using [phpstan v1.10.5] https://megalinter.io/7411316caa2897b3839af1834e5fd8ebd5b7b81e/descriptors/php_phpstan
  - MegaLinter key: [PHP_PHPSTAN]
  - Rules config: [phpstan.neon.dist]
  - Number of files analyzed: [1]
  --Error detail:
  Error: Method Prql\Compiler\Compiler::compile() has invalid return type Prql\Compiler\Result.
  Error: Parameter $options of method Prql\Compiler\Compiler::compile() has invalid type Prql\Compiler\Options.
  Error: Method Prql\Compiler\Compiler::prqlToPL() has invalid return type Prql\Compiler\Result.
  Error: Method Prql\Compiler\Compiler::plToRQ() has invalid return type Prql\Compiler\Result.
  Error: Method Prql\Compiler\Compiler::rqToSQL() has invalid return type Prql\Compiler\Result.
  Error: Parameter $options of method Prql\Compiler\Compiler::rqToSQL() has invalid type Prql\Compiler\Options.
  Error: Parameter $options of method Prql\Compiler\Compiler::optionsInit() has invalid type Prql\Compiler\Options.
  Error: Instantiated class Prql\Compiler\Options not found.
  Error: Method Prql\Compiler\Compiler::convertResult() has invalid return type Prql\Compiler\Result.
  Error: Instantiated class Prql\Compiler\Result not found.
  Error: Method Prql\Compiler\Compiler::convertMessage() has invalid return type Prql\Compiler\Message.
  Error: Instantiated class Prql\Compiler\Message not found.
  Error: Access to constant Error on an unknown class Prql\Compiler\MessageKind.
  Error: Access to constant Warning on an unknown class Prql\Compiler\MessageKind.
  Error: Access to constant Lint on an unknown class Prql\Compiler\MessageKind.
  Error: Method Prql\Compiler\Compiler::convertSpan() has invalid return type Prql\Compiler\Span.
  Error: Instantiated class Prql\Compiler\Span not found.
  Error: Method Prql\Compiler\Compiler::convertLocation() has invalid return type Prql\Compiler\SourceLocation.
  Error: Instantiated class Prql\Compiler\SourceLocation not found.
```

Should we disable phpstan as this is obviously confusing?
Or does it make sense to always inspect the entire repository?

File: .mega-linter.yaml (hunk 1 above)
Comment (nvuillam) [reply .mega-linter.yaml]: @eitsupi no problem :)
i think it should work, but usually it's easier to read and maintain to define one line by argument :)

```yaml
PHP_PHPCS_ARGUMENTS:
  - --standard=PSR12
```

Comment (eitsupi) [reply .mega-linter.yaml]: Thank you for your immediate reply!
I will update like that.

**Comment by max-sixty** [comment]
> Should we disable phpstan as this is obviously confusing?
> Or does it make sense to always inspect the entire repository?

Yeah, if it's producing things that are obviously wrong, we can disable and leave a comment pointing to this issue. We don't have to disable everything that produces any false positive, but if it's spewing them and not producing helpful messages, +1 to disable...

**Review by max-sixty — APPROVED** [review]

File: .mega-linter.yaml (hunk 1 above)
Comment (nvuillam) [reply .mega-linter.yaml]: Thank you for using MegaLinter :)
Don't hesitate to post about it if you like it 👼

File: prql-php/src/Compiler.php [hunk 2]
```diff
@@ -39,17 +39,27 @@ final class Compiler
...
             $library = $lib_path."\libprql_lib.dll";
+            // phpcs:enable
         } elseif (PHP_OS_FAMILY === 'Darwin') {
+            // phpcs:disable PSR12.Operators.OperatorSpacing
             $library = $lib_path.'/libprql_lib.dylib';
+            // phpcs:enable
         } else {
+            // phpcs:disable PSR12.Operators.OperatorSpacing
             $library = $lib_path.'/libprql_lib.so';
+            // phpcs:enable
```
Comment (vanillajonathan) [inline prql-php/src/Compiler.php]:
```suggestion
            $lib_path = __DIR__ . '/../lib';
        }

        $header = $lib_path . '/libprql_lib.h';

        if (PHP_OS_FAMILY === 'Windows') {
            $library = $lib_path . "\libprql_lib.dll";
        } elseif (PHP_OS_FAMILY === 'Darwin') {
            $library = $lib_path . '/libprql_lib.dylib';
        } else {
            $library = $lib_path . '/libprql_lib.so';
```

**Comment by vanillajonathan** [comment]
I think psalm not knowing about `\FFI` is a false positive. I don't know why it doesn't know about it. [FFI](https://www.php.net/manual/en/class.ffi.php) should exist. The `\` (slash) indicates it is outside of the Prql namespace.

Please use spaces around the dot when concatenating strings.

It looks to me like the complaints by PHPStan has to do with it not finding the classes that are referenced.
https://phpstan.org/user-guide/discovering-symbols

PHPStan should be able to find the references classes since the autoloader is defined in the `composer.json` file. However it might be that PHPStan cannot find the the `composer.json` file since it is not in the root directory? Maybe there is some way to configure Megalinter to tell PHPStan that it should use prql-php/ as the working directory?
It would be good to try run PHPStan on the local machine in the `prql-php` directory to see if it finds all the referenced classes then in order to figure it out.

**Comment by eitsupi** [comment]
@vanillajonathan Thanks for taking a look at this!
Unfortunately I could not figure out how to prevent false positives from phpstan.
So please let me disable this for now.

What remains are the following 8 errors by phpcs, do these errors make sense?
If it is a false positive, I think this rule needs to be disabled.

```log
  ----------------------------------------------------------------------
  FOUND 8 ERRORS AND 1 WARNING AFFECTING 9 LINES
  ----------------------------------------------------------------------
    34 | WARNING | Property name "$_ffi" should not be prefixed with an
       |         | underscore to indicate visibility
   162 | ERROR   | Method name "_optionsInit" must not be prefixed with
       |         | an underscore to indicate visibility
   181 | ERROR   | Method name "_optionsDestroy" must not be prefixed
       |         | with an underscore to indicate visibility
   189 | ERROR   | Method name "_convertResult" must not be prefixed
       |         | with an underscore to indicate visibility
   207 | ERROR   | Method name "_convertMessage" must not be prefixed
       |         | with an underscore to indicate visibility
   231 | ERROR   | Method name "_convertSpan" must not be prefixed with
       |         | an underscore to indicate visibility
   243 | ERROR   | Method name "_convertLocation" must not be prefixed
       |         | with an underscore to indicate visibility
   258 | ERROR   | Method name "_convertNullableString" must not be
       |         | prefixed with an underscore to indicate visibility
   267 | ERROR   | Method name "_convertString" must not be prefixed
       |         | with an underscore to indicate visibility
  ----------------------------------------------------------------------
```

**Comment by vanillajonathan** [comment]
> @vanillajonathan Thanks for taking a look at this! Unfortunately I could not figure out how to prevent false positives from phpstan. So please let me disable this for now.
> 
> What remains are the following 8 errors by phpcs, do these errors make sense? If it is a false positive, I think this rule needs to be disabled.
> 
> ```
>   ----------------------------------------------------------------------
>   FOUND 8 ERRORS AND 1 WARNING AFFECTING 9 LINES
>   ----------------------------------------------------------------------
>     34 | WARNING | Property name "$_ffi" should not be prefixed with an
>        |         | underscore to indicate visibility
>    162 | ERROR   | Method name "_optionsInit" must not be prefixed with
>        |         | an underscore to indicate visibility
>    181 | ERROR   | Method name "_optionsDestroy" must not be prefixed
>        |         | with an underscore to indicate visibility
>    189 | ERROR   | Method name "_convertResult" must not be prefixed
>        |         | with an underscore to indicate visibility
>    207 | ERROR   | Method name "_convertMessage" must not be prefixed
>        |         | with an underscore to indicate visibility
>    231 | ERROR   | Method name "_convertSpan" must not be prefixed with
>        |         | an underscore to indicate visibility
>    243 | ERROR   | Method name "_convertLocation" must not be prefixed
>        |         | with an underscore to indicate visibility
>    258 | ERROR   | Method name "_convertNullableString" must not be
>        |         | prefixed with an underscore to indicate visibility
>    267 | ERROR   | Method name "_convertString" must not be prefixed
>        |         | with an underscore to indicate visibility
>   ----------------------------------------------------------------------
> ```

I fixed those errors and warnings with PR #2096.

**Comment by eitsupi** [comment]
> I fixed those errors and warnings with PR #2096.

I've already included these fixes in this branch (thanks!).

But I don't think these errors have been fixed.
https://github.com/PRQL/prql/actions/runs/4389422640/jobs/7686949039#step:4:365

**Comment by vanillajonathan** [comment]
Oh, now we have conflicting rule sets applied. I did my PR by running `phpcs src` which uses the PEAR coding standard by default, and you pass `--standard=PSR12` argument to use the PSR12 standard.
I wasn't aware of this until now, and I think using PSR12 is probably better in a modern project.

**Comment by eitsupi** [comment]
Ok, let's undo your change, since the README states to use `--standard=PSR12`.
https://github.com/PRQL/prql/blob/ed1e896ecbb1c5e18f0f1286639118e3a77b0100/prql-php/README.md?plain=1#L65-L69

**Comment by vanillajonathan** [comment]
@eitsupi Yeah, I think PSR-12 is better, so I have a PR to fix that. #2106

**Comment by eitsupi** [comment]
Now that the PHP checks are all green, let's merge this.
Thanks @vanillajonathan!

```
 +----SUMMARY----+------------+---------------+-------+-------+--------+--------------+
| Descriptor    | Linter     | Mode          | Files | Fixed | Errors | Elapsed time |
+---------------+------------+---------------+-------+-------+--------+--------------+
| ◬ COPYPASTE   | jscpd      | project       |   n/a |       |      8 |       27.95s |
| ✅ PHP        | phpcs      | list_of_files |     1 |       |      0 |        0.16s |
| ✅ PHP        | phplint    | list_of_files |     1 |       |      0 |        0.06s |
| ◬ REPOSITORY  | checkov    | project       |   n/a |       |     22 |       19.35s |
| ◬ REPOSITORY  | devskim    | project       |   n/a |       |      1 |        1.32s |
| ✅ REPOSITORY | dustilock  | project       |   n/a |       |      0 |        1.63s |
| ✅ REPOSITORY | gitleaks   | project       |   n/a |       |      0 |        6.37s |
| ✅ REPOSITORY | git_diff   | project       |   n/a |       |      0 |        0.05s |
| ✅ REPOSITORY | secretlint | project       |   n/a |       |      0 |        4.75s |
| ✅ REPOSITORY | syft       | project       |   n/a |       |      0 |        1.07s |
| ◬ REPOSITORY  | trivy      | project       |   n/a |       |      1 |        6.37s |
| ✅ SPELL      | misspell   | list_of_files |     2 |       |      0 |        0.07s |
| ✅ YAML       | prettier   | list_of_files |     1 |       |      0 |        0.54s |
| ✅ YAML       | v8r        | list_of_files |     1 |       |      0 |        1.95s |
| ✅ YAML       | yamllint   | list_of_files |     1 |       |      0 |        0.26s |
+---------------+------------+---------------+-------+-------+--------+--------------+
```

**Comment by vanillajonathan** [comment]
PHPStan works locally.

    composer require --dev phpstan/phpstan
    ./vendor/bin/phpstan analyze src

It has no errors with the default level 0 check, but with level 1 and check it gets errors. PHPStan have levels up to 9.

    ./vendor/bin/phpstan analyze --level=9 src

Some of the errors are because some of the private methods do not have parameter types and return types defined. Other errors are because PHPStan does not know about our methods and fields defined in our C library that we invoke through the FFI object since the FFI object does not have those defined.

**Pull request merged**